// Copyright (c) 2016-2017, Nefeli Networks, Inc.
// Copyright (c) 2017, The Regents of the University of California.
// All rights reserved.
//
// Redistribution and use in source and binary forms, with or without
// modification, are permitted provided that the following conditions are met:
//
// * Redistributions of source code must retain the above copyright notice, this
// list of conditions and the following disclaimer.
//
// * Redistributions in binary form must reproduce the above copyright notice,
// this list of conditions and the following disclaimer in the documentation
// and/or other materials provided with the distribution.
//
// * Neither the names of the copyright holders nor the names of their
// contributors may be used to endorse or promote products derived from this
// software without specific prior written permission.
//
// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
// POSSIBILITY OF SUCH DAMAGE.

syntax = "proto3";

package ddsketch.pb;

/**
 * The Module DDSketch
 */
 message DDSketchArg {
  double accuracy = 1;
  uint64 max_bucket_number = 2;
  uint64 offset = 3;
}


/**
* Clears the captured data.
*/
message DDSketchCommandEmptyArg{}


/**
* Returns the status of the collected data.
*/
message DDSketchCommandGetStatArg{}


/**
* Contains the status of the collected data.
*/
message DDSketchCommandGetStatResponse{
  uint64 packet_number = 1;
  uint64 bucket_number = 2;
  uint64 max_bucket_number = 3;
  double accuracy = 4;
  double lambda = 5;
}

/**
* Returns the status of the collected data.
*/
message DDSketchCommandGetContentArg{}


/**
* Contains the status of the collected data.
*/
message DDSketchCommandGetContentResponse{
  message Bucket{
    uint64 index = 1;
    uint64 counter = 2;
  }
  repeated Bucket content = 1;
}


/**
* Returns the wanted quantile.
*/
message DDSketchCommandGetQuantileArg{
  uint32 quantile = 1;
}


/**
* Contains the wanted quantile.
*/
message DDSketchCommandGetQuantileResponse{
  uint32 quantile = 1;
}